### PR TITLE
DYN-3853: Similar DS code blocks give inconsistent results

### PR DESCRIPTION
**Pull request by sm6srw**
### Purpose

[DYN-3853](https://jira.autodesk.com/browse/DYN-3853) - Similar DS code blocks give inconsistent results

This pull request does:
* check for `IdentifierListNode` before `ArrayNameNode` in utility methods.
* a small refactoring of utility methods to make them easier to read
* add tests covering the failing case.

#### Background

A code block like: 

```
t3 = List.RemoveItemAtIndex(l2,
List.Cycle(l1, num)@L1<1>);
```

are expected to generate the same code as a code block like:

```
t1 = List.Cycle(l1, num);
t3 = List.RemoveItemAtIndex(l2,
t1@L1<1>);
```

The code generated for the first code block was wrong:

```
==============Start Node==============
[a.0.9140]push %t_1143_85709d54aab04c8196338b947db93ae5
[a.0.9141]pushlevel @0
[a.0.9142]popguides 0
[a.0.9143]push %t_1147_85709d54aab04c8196338b947db93ae5
[a.0.9144]pushlevel @0
[a.0.9145]pushguide 1
[a.0.9146]popguides 1
[a.0.9147]push 0[depth]
[a.0.9148]callr RemoveItemAtIndex
[a.0.9149]push _rx
[a.0.9150]pop %t_1148_85709d54aab04c8196338b947db93ae5
[a.0.9151]dep 686[ExprUID] False[SSA]
==============End Node==============
```

compared to the second code block that is correct:

```
a.0.5494]push %t_623_85709d54aab04c8196338b947db93ae5
[a.0.5495]pushlevel @0
[a.0.5496]popguides 0
[a.0.5497]push %t_624_85709d54aab04c8196338b947db93ae5
[a.0.5498]pushlevel @L1
[a.0.5499]pushguide 1
[a.0.5500]popguides 1
[a.0.5501]push 0[depth]
[a.0.5502]callr RemoveItemAtIndex
[a.0.5503]push _rx
[a.0.5504]pop %t_625_85709d54aab04c8196338b947db93ae5
[a.0.5505]dep 602[ExprUID] False[SSA]
```

(the difference is the second `pushlevel @0` that should be `pushlevel @L1`)

The problem was that we failed to pick up the node at level information correctly when the at level information was applied directly to an expression.  An expression is a `IdentifierListNode` (that is derived from  `ArrayNameNode`) internally. Some utility methods for at level management where incorrectly checking if nodes where  a `ArrayNameNode` before checking if they where a `IdentifierListNode` and that caused the code that where supposed to take care of this case to never execute.

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [X] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### CODE REVIEW

File: src/Engine/ProtoAssociative/CodeGen.cs [hunk 1]
```diff
@@ -1861,96 +1861,95 @@ public ProtoCore.DSASM.ProcedureNode GetProcedureFromInstance(int classScope, Pr
...
+        /// This helper function extracts the at level data from the 
+        /// AST node
+        /// </summary>
+        /// <param name="node"></param>
+        /// <returns>At level data</returns>
         private AtLevelNode GetAtLevel(AssociativeNode node)
         {
-            var n1 = node as ArrayNameNode;
-            if (n1 != null)
+            if (node is IdentifierListNode identifierListNode)
```
Comment (pinzart90) [inline src/Engine/ProtoAssociative/CodeGen.cs:1893]: can you add a note (in the code) here that the order matters ?
First check the most derived class IdentifierListNode, then the base class ArrayNameNode

**Review by pinzart90 — APPROVED** [review]
LGTM

File: test/DynamoCoreTests/ListAtLevelTests.cs [hunk 2]
```diff
@@ -73,5 +73,12 @@ public void TestUseListStructureAndLacing2()
             RunModel(testFilePath);
             AssertPreviewValue("ae89387d-77b8-49e2-b59a-26a4b0a2ae9b", new[] { new[] { "3a", "7b" }, new[] { "11c", "15c" } });
         }
+        [Test]
+        public void TestUseListAtLevelOnFunctionCall()
+        {
+            string testFilePath = Path.Combine(listTestFolder, "testfunctioncall.dyn");
+            RunModel(testFilePath);
+            AssertPreviewValue("b3cd769e-0f9d-4b4f-ac86-53a07edaba24", new[] { new[] { new[] { 5, 6 } }, new[] { new[] { 5, 6 }, new[] { 7, 8 } }, new[] { new[] { 5, 6 }, new[] { 7, 8 } }, new[] { new[] { 5, 6 }, new[] { 7, 8 } } });
+        }
```
Comment (aparajit-pratap) [inline test/DynamoCoreTests/ListAtLevelTests.cs:82]: Can we also have a pure language level test for this here: https://github.com/DynamoDS/Dynamo/blob/master/test/Engine/ProtoTest/AtLevel/AtLevelTest.cs?

**Review by aparajit-pratap — COMMENTED** [review]
Good catch!

File: test/DynamoCoreTests/ListAtLevelTests.cs (hunk 2 above)
Comment (sm6srw) [reply test/DynamoCoreTests/ListAtLevelTests.cs:82]: Thanks! I didn't notice those. Will do.

**Review by aparajit-pratap — APPROVED** [review]

**Pull request merged**